### PR TITLE
feat(grid): base implementation

### PR DESCRIPTION
**Pull request by bmdalex**
# Grid

This PR will introduce the **Grid** component with the following props: 
- `columns=[number|string]`
- `rows=[number|string]`

The Grid will specify the layout of the elements it encapsulates as `children` or in the `content` prop.

See also:
- #11 - for description of what we're trying to accomplish
- https://learncssgrid.com/ and
https://css-tricks.com/snippets/css/complete-guide-grid/ - for CSS grid layout ramp-up

### TODO

- [x] Conformance test
- [x] Minimal doc site example
- [ ] Stardust base theme
- [x] Teams Light theme
- [ ] Teams Dark theme
- [ ] Teams Contrast theme
- [ ] Confirm RTL usage
- [ ] [W3 accessibility](https://www.w3.org/standards/webdesign/accessibility) check
- [ ] [Stardust accessibility](https://github.com/stardust-ui/accessibility) check
- [ ] Update glossary props table
- [x] Update the CHANGELOG.md

# API Proposal

## `columns`

We can specify a certain amount of columns:

columncount

or the explicit columns for a grid:

columnexplicit

```jsx
<Grid columns="7">
  <Image fluid src="img1.png" />
  <Image fluid src="img2.png" />
  <Image fluid src="img3.png" />
  ...
</Grid>
```

and

```jsx
<Grid columns="repeat(3, 1fr) 2fr 2fr 110px 14rem 50px 20%">
  <Image fluid src="img1.png" />
  <Image fluid src="img2.png" />
  <Image fluid src="img3.png" />
  ...
</Grid>
```

generate:

```html
<div class="ui-grid a b c d e f k">
  <img src="img1.png" class="ui-image h i b j" />
  <img src="img2.png" class="ui-image h i b j" />
  <img src="img3.png" class="ui-image h i b j" />
  ...
</div>
```

## `rows`

We can specify a certain amount of rows:

rowcount

or the explicit rows for a grid:

rowexplicit

```jsx
<Grid rows="2">
  <Image fluid src="img1.png" />
  <Image fluid src="img2.png" />
  <Image fluid src="img3.png" />
  ...
</Grid>
```

and

```jsx
<Grid rows="2fr repeat(2, 1fr)">
  <Image fluid src="img1.png" />
  <Image fluid src="img2.png" />
  <Image fluid src="img3.png" />
  ...
</Grid>
```

generate:

```html
<div class="ui-grid a b c d e f k">
  <img src="img1.png" class="ui-image h i b j" />
  <img src="img2.png" class="ui-image h i b j" />
  <img src="img3.png" class="ui-image h i b j" />
  ...
</div>
```

### CODE REVIEW

**Comment by gaui** [comment]
What are those extra letter classes? `a b c d e f k`

**Comment by Bugaa92** [comment]
@kuzhelov no, this new Grid version does not have the concept of rows and columns as these are controlled with props; see proposal #11.

Therefore, all we have are grid items (that can be any element as we do not have any constraints there); in the future we're probably going to add a `GridItem` component for controlling styles of individual items

**Comment by Bugaa92** [comment]
@gaui Fella related classes that add the styles internally; if it's confusing I'm gonna delete them from the example as they are irrelevant there

**Comment by Bugaa92** [comment]
@levithomason removed `itemSize` from the API and adjusted default gap and padding; this is ready for review now and can probably be merged as is

**Review by levithomason — CHANGES_REQUESTED** [review]
Per chat, let's only merge rows / columns for now.  These props hit our use cases.

The `itemSize` can be discussed in more detail later on subsequent PRs.  It has some implications that will need collaboratively worked out before merging.

File: docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+const images = imageNames.map((name, index) => (
+  <Image key={`${name}-${index}`} fluid src={`public/images/avatar/large/${name}`} />
+))
+
+const GridExample = () => (
+  <div>
+    <Grid content={images} />
+  </div>
+)
```
Comment (levithomason) [inline docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx]: Wrapping `div` is not required, this can be reduced to:

```jsx
const GridExample = () => <Grid content={images} />
```

File: docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  'elyse.png',
+  'helen.jpg',
+  'jenny.jpg',
+  'joe.jpg',
+  'justen.jpg',
+]
+
+const images = imageNames.map((name, index) => (
+  <Image key={`${name}-${index}`} fluid src={`public/images/avatar/large/${name}`} />
+))
```
Comment (levithomason) [inline docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx]: Given the amount of code required to abstract and loop over the images, it would be less code and easier for users to scan if we inline the images:

```jsx
<Image fluid src="public/images/avatar/large/ade.jpg" />
<Image fluid src="public/images/avatar/large/chris.jpg" />
<Image fluid src="public/images/avatar/large/christian.jpg" />
<Image fluid src="public/images/avatar/large/daniel.jpg" />
<Image fluid src="public/images/avatar/large/elliot.jpg" />
<Image fluid src="public/images/avatar/large/elyse.png" />
<Image fluid src="public/images/avatar/large/helen.jpg" />
<Image fluid src="public/images/avatar/large/jenny.jpg" />
<Image fluid src="public/images/avatar/large/joe.jpg" />
<Image fluid src="public/images/avatar/large/justen.jpg" />
```

Comment (bmdalex) [reply docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx]: @levithomason 
I have 20+ images in other grid examples to make it clear to users how the grid looks like ; duplicating 20+ lines in other places doesn't really make sense; do we want to change it only for these 2 examples (`GridExample.shorthand.tsx` and `GridExample.tsx`) ?

File: src/components/Grid/Grid.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    columns: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+
+    /** Shorthand for primary content. */
+    content: customPropTypes.every([
+      customPropTypes.disallow(['children']),
+      PropTypes.oneOfType([
+        PropTypes.arrayOf(customPropTypes.itemShorthand),
+        customPropTypes.itemShorthand,
+      ]),
+    ]),
```
Comment (levithomason) [inline src/components/Grid/Grid.tsx:36]: In the case of a Grid component, I'm not sure `content` makes sense.  Perhaps here we only support children?

Why? There is no processing to be done on the content by Grid.  It is also an "invisible" component in that it only formats children, but doesn't render its own subcomponents.

Thoughts?

File: src/themes/teams/componentStyles.ts [hunk 4]
```diff
@@ -11,24 +10,21 @@ export { default as ChatMessage } from './components/Chat/chatMessageStyles'
...
 
 export { default as Label } from './components/Label/labelStyles'
-
 export { default as Layout } from './components/Layout/layoutStyles'
-
 export { default as List } from './components/List/listStyles'
 export { default as ListItem } from './components/List/listItemStyles'
 
 export { default as Menu } from './components/Menu/menuStyles'
-
```
Comment (levithomason) [inline src/themes/teams/componentStyles.ts]: Let's avoid unnecessary whitespace changes.  It will make the PR smaller and prevent us from flip flopping on code style updates over time.

A good way to propose and implement style fixes is to make a PR introducing auto fixable lint rules or prettier config.  This way, code style is automatic and enforced.

File: docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx (hunk 2 above)
Comment (levithomason) [reply docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx]: The criteria for examples is concise and readable.  We want them to fit on the minimal number of lines so that the can be easily viewed inline in the doc site.  In cases where abstractions significantly reduce the amount code, then they are helpful and we should use them.

In the example here, given any number of images, the abstracted code will require more lines of code since each image path in the array already requires a line of code.  Any lines required for abstraction are not needed.

In examples where the array of images is used more than once, then we'll certainly saves lines by abstracting.  We should use an abstraction for those.  However, you might consider removing the `map` code and just creating an array of images once:

```jsx
const images = [
  <Image key="ade" fluid src="public/images/avatar/large/ade.jpg" />,
  <Image key="chris" fluid src="public/images/avatar/large/chris.jpg" />,
  <Image key="christian" fluid src="public/images/avatar/large/christian.jpg" />,
  <Image key="daniel" fluid src="public/images/avatar/large/daniel.jpg" />,
  <Image key="elliot" fluid src="public/images/avatar/large/elliot.jpg" />,
  <Image key="elyse" fluid src="public/images/avatar/large/elyse.png" />,
  <Image key="helen" fluid src="public/images/avatar/large/helen.jpg" />,
  <Image key="jenny" fluid src="public/images/avatar/large/jenny.jpg" />,
  <Image key="joe" fluid src="public/images/avatar/large/joe.jpg" />,
  <Image key="justen" fluid src="public/images/avatar/large/justen.jpg" />,
]
```

File: src/components/Grid/Grid.tsx (hunk 3 above)
Comment (bmdalex) [reply src/components/Grid/Grid.tsx:36]: I agree with the reasons, but we have following things preventing us to do so:
- all components should support shorthand version; removing `content` breaks that pattern basically; it'd be the first component without shorthand support, right?
- we were discussing introducing `GridItem` component for controlling individual items; maybe it'll make sense at that point

Comment (levithomason) [reply src/components/Grid/Grid.tsx:36]: Yes, good points.  We probably should not make those types of decisions in this PR.  Let's keep it as is and we can discuss it in another.  This would be a framework level design change.

File: src/themes/teams/componentStyles.ts (hunk 4 above)
Comment (bmdalex) [reply src/themes/teams/componentStyles.ts]: I was just removing some random new lines, at least grouping by components with same letter; ok, I'll add my change only

File: docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx (hunk 1 above)
Comment (bmdalex) [reply docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx]: done

File: docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx (hunk 2 above)
Comment (bmdalex) [reply docs/src/examples/components/Grid/Types/GridExample.shorthand.tsx]: done

File: src/themes/teams/componentStyles.ts (hunk 4 above)
Comment (bmdalex) [reply src/themes/teams/componentStyles.ts]: done

**Comment by Bugaa92** [comment]
@levithomason I see some visual tests failing but I cannot login to this `screener ` thingie to see what's up

**Comment by levithomason** [comment]
Send me the email you'd like to use on Screener privately and I'll add you.

**Comment by bmdalex** [comment]
@levithomason solved the screener issue with help from Miro. Everything is passing and I think I addressed all comments

**Review by levithomason — APPROVED** [review]
